### ISSUE TITLE
(Feature Request) Finer control over which lanes are disabled during lane switches

### ISSUE THREAD

**Issue by percyqaz** [InventiveRhythm/fluXis#107]
TLDR: Currently lane switches specify a new lane count for a map, and the game automatically decides which of the lanes should now be visible

What if mappers could specify which lanes are hidden/shown by a lane switch?

Video of how it could look (idk how to embed)
https://github.com/user-attachments/assets/c4ee8bc0-eb33-4f10-98c6-f2196cc7eb8e

# Considerations

Chances are you've already thought about this given the design of the existing code

## Implementation
The JSON for lane switches could go from something like this:
`{"time":3868.7271728515625,"count":4,"speed":0.0,"easing":13}`
To something like this:
`{"time":3868.7271728515625,"lanes":[2,3,5,6],"speed":0.0,"easing":13}`

Columns start at 1, with 1 = the leftmost column
Invalid numbers in the list could be discarded or ignored

Open to discussion on this, it could also be an array of booleans

To actually render the lane switches, conveniently the code is already structured with a hard-coded `SWITCH_VISIBILITY` array defining what columns are visible when the base keycount is N and the current keycount is K (K less then or equal to N)
Instead of using this array, the array comes from the most recent lane switch event

~~I'm assuming this same code is used in the editor to highlight the disabled lanes red, but I haven't checked - If so that change will also come pretty easily~~
Edit: Here is the editor correctly handling it - it came for free with changing it for gameplay
![image](https://github.com/user-attachments/assets/a130d0de-0f70-43a0-9347-d3e06e2253de)

## Backwards compatibility
If not present, the value of the new `lanes` property on lane switches can be calculated from the map's keycount + the existing `count` property, by keeping that hard-coded SWITCH_VISIBILITY array somewhere

Open to discussion on this

## Interaction with mods
#94 can be fixed by making the Mirror mod implement `IApplicableToEvents` and then mirror the lanes on all lane switch events
This way the active lanes match the place where mirrored notes are

Similarly if any future mods rearrange the columns in other ways, as long as the same reordering is applied to lane switch lanes the mod will work correctly

## Gameplay considerations
My thought is that enabled lanes should behave exactly like the original lanes in the map as if all columns were there, for example:

- In an 8k map, if only the left 4 keys are enabled, the binds to play those columns are still the left 4 keys which means a map can encourage a one-handed section (this does open up debates about cheesing with binds and stuff, but I think it would be fun)
- In that same map with the mirror mod applied, you play one-handed with your right hand instead
- If your skin uses lane-specific colours they should remain the colours for the appropriate lanes

I have not investigated fully if this is already the case or any other technical concerns with skinning this could create, so please let me know

## UI for editor
Open to discussion on this, I'm picturing a row of checkboxes and you can just uncheck certain lanes
A first draft of the feature may not implement this and leave it to editing .ffx files manually

@flustix I am willing to try implementing the majority or all of this feature myself as a PR, with your approval if the discussion points are resolved

**Comment by percyqaz** [comment]
So I had a crack at hacking it in just to see how it feels
https://github.com/user-attachments/assets/c4ee8bc0-eb33-4f10-98c6-f2196cc7eb8e

It's not quite as good as I imagined it and there will be some practical concerns:
- Needing to telegraph lane switches really well because 1 lane morphing into another can be easily missed
  Follow-up idea: telegraph lane switches by moving the playfield on the screen
- Maybe there is a need to keep the keybind overlay hints on all the time when a map has lane switches
- Changes in column widths and hitpositions can be a bit weird even with lane switches currently

So maybe the game isn't ready for this yet
But there you go at least it's an idea

**Comment by flustix** [comment]
This has already been suggested.
And rejected.

Simply put, consistency. It will confuse players (especially new ones) if every map just has a different way it does lane switches.

**Comment by percyqaz** [comment]
> This has already been suggested.

Sorry, couldn't find a duplicate anywhere

> Simply put, consistency.

Makes sense to me, I could also take it or leave it after seeing it wasn't as cool as it was in my head
At least I got to satisfy my curiosity :)
